File: apps/agent/src/lib/graph.ts
import { ObjectId, type Collection } from "mongodb";
import { getDb } from "./mongo.js";

export type EntityKind = "person" | "project" | "topic";

export interface EntityRecord {
  _id?: ObjectId;
  name: string;             // canonical display name
  key: string;              // lowercased + de-noised key for matching
  kind: EntityKind;
  role?: string;            // one-line description (e.g. "Director of Eng, Lantern team")
  mentions: number;         // count of chunks mentioning this entity
  firstSeen?: string;       // ISO date
  lastSeen?: string;        // ISO date
  chunkIds: string[];       // sample of chunks where mentioned (cap at 12)
  // sparkline series: mentions per day across the corpus window (sorted by date)
  series: Array<{ date: string; count: number }>;
  createdAt: Date;
  updatedAt: Date;
}

export interface RelationRecord {
  _id?: ObjectId;
  from: string;             // entity key (owner)
  to: string;               // entity key (recipient)
  kind: "owes" | "works_with" | "manages" | "discusses";
  evidence: string;         // verbatim phrase or short summary
  chunkId?: string;
  date?: string;
  createdAt: Date;
}

export async function entitiesCol(): Promise<Collection<EntityRecord>> {
  const db = await getDb();
  return db.collection<EntityRecord>("entities");
}

export async function relationsCol(): Promise<Collection<RelationRecord>> {
  const db = await getDb();
  return db.collection<RelationRecord>("relations");
}

export async function upsertEntity(rec: Omit<EntityRecord, "_id" | "createdAt" | "updatedAt">): Promise<void> {
  const col = await entitiesCol();
  const now = new Date();
  await col.updateOne(
    { key: rec.key, kind: rec.kind },
    {
      $set: {
        name: rec.name,
        role: rec.role,
        mentions: rec.mentions,
        firstSeen: rec.firstSeen,
        lastSeen: rec.lastSeen,
        chunkIds: rec.chunkIds.slice(0, 12),
        series: rec.series,
        updatedAt: now,
      },
      $setOnInsert: {
        key: rec.key,
        kind: rec.kind,
        createdAt: now,
      },
    },
    { upsert: true },
  );
}

export async function insertRelation(rec: Omit<RelationRecord, "_id" | "createdAt">): Promise<void> {
  const col = await relationsCol();
  // de-dup on (from, to, kind, chunkId)
  const filter: Record<string, unknown> = { from: rec.from, to: rec.to, kind: rec.kind };
  filter["chunkId"] = rec.chunkId ?? { $exists: false };
  await col.updateOne(
    filter,
    { $setOnInsert: { ...rec, createdAt: new Date() } },
    { upsert: true },
  );
}

export async function listEntities(opts: { kind?: EntityKind; limit?: number } = {}): Promise<EntityRecord[]> {
  const col = await entitiesCol();
  const filter: Record<string, unknown> = {};
  if (opts.kind) filter["kind"] = opts.kind;
  return col.find(filter, { limit: opts.limit ?? 200, sort: { mentions: -1 } }).toArray();
}

export async function listRelations(opts: { kind?: RelationRecord["kind"]; limit?: number } = {}): Promise<RelationRecord[]> {
  const col = await relationsCol();
  const filter: Record<string, unknown> = {};
  if (opts.kind) filter["kind"] = opts.kind;
  return col.find(filter, { limit: opts.limit ?? 200, sort: { createdAt: -1 } }).toArray();
}

export async function graphStats(): Promise<{
  entities: { person: number; project: number; topic: number };
  relations: number;
}> {
  const ents = await entitiesCol();
  const rels = await relationsCol();
  const [people, projects, topics, relCount] = await Promise.all([
    ents.countDocuments({ kind: "person" }),
    ents.countDocuments({ kind: "project" }),
    ents.countDocuments({ kind: "topic" }),
    rels.countDocuments({}),
  ]);
  return { entities: { person: people, project: projects, topic: topics }, relations: relCount };
}

export async function clearGraph(): Promise<void> {
  const [ents, rels] = await Promise.all([entitiesCol(), relationsCol()]);
  await Promise.all([ents.deleteMany({}), rels.deleteMany({})]);
}

export function publicEntity(e: EntityRecord): Record<string, unknown> {
  return {
    id: e._id?.toString() ?? "",
    name: e.name,
    key: e.key,
    kind: e.kind,
    role: e.role ?? null,
    mentions: e.mentions,
    firstSeen: e.firstSeen ?? null,
    lastSeen: e.lastSeen ?? null,
    chunkIds: e.chunkIds,
    series: e.series,
  };
}

export function publicRelation(r: RelationRecord): Record<string, unknown> {
  return {
    id: r._id?.toString() ?? "",
    from: r.from,
    to: r.to,
    kind: r.kind,
    evidence: r.evidence,
    chunkId: r.chunkId ?? null,
    date: r.date ?? null,
  };
}

export function entityKey(name: string): string {
  return name.trim().toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "");
}

File: apps/agent-py/app/lib/graph.py
"""Memory graph store — port of apps/agent/src/lib/graph.ts."""
from __future__ import annotations

import re
from datetime import datetime, timezone

from app.db.mongo import collection


def entity_key(name: str) -> str:
    return re.sub(r"^-|-$", "", re.sub(r"[^a-z0-9]+", "-", name.strip().lower()))


def entities_col():
    return collection("entities")


def relations_col():
    return collection("relations")


async def upsert_entity(rec: dict) -> None:
    now = datetime.now(timezone.utc)
    await entities_col().update_one(
        {"key": rec["key"], "kind": rec["kind"]},
        {
            "$set": {
                "name": rec["name"], "role": rec.get("role"), "mentions": rec["mentions"],
                "firstSeen": rec.get("firstSeen"), "lastSeen": rec.get("lastSeen"),
                "chunkIds": rec["chunkIds"][:12], "series": rec["series"], "updatedAt": now,
            },
            "$setOnInsert": {"key": rec["key"], "kind": rec["kind"], "createdAt": now},
        },
        upsert=True,
    )


async def insert_relation(rec: dict) -> None:
    flt = {"from": rec["from"], "to": rec["to"], "kind": rec["kind"],
           "chunkId": rec.get("chunkId", {"$exists": False})}
    await relations_col().update_one(
        flt, {"$setOnInsert": {**rec, "createdAt": datetime.now(timezone.utc)}}, upsert=True
    )


async def list_entities(kind: str | None = None, limit: int = 200) -> list[dict]:
    flt = {"kind": kind} if kind else {}
    cur = entities_col().find(flt, limit=limit, sort=[("mentions", -1)])
    return await cur.to_list(length=None)


async def list_relations(limit: int = 200) -> list[dict]:
    cur = relations_col().find({}, limit=limit, sort=[("createdAt", -1)])
    return await cur.to_list(length=None)


async def graph_stats() -> dict:
    ents, rels = entities_col(), relations_col()
    return {
        "entities": {
            "person": await ents.count_documents({"kind": "person"}),
            "project": await ents.count_documents({"kind": "project"}),
            "topic": await ents.count_documents({"kind": "topic"}),
        },
        "relations": await rels.count_documents({}),
    }


async def clear_graph() -> None:
    await entities_col().delete_many({})
    await relations_col().delete_many({})


def public_entity(e: dict) -> dict:
    return {
        "id": str(e.get("_id", "")), "name": e["name"], "key": e["key"], "kind": e["kind"],
        "role": e.get("role"), "mentions": e.get("mentions", 0),
        "firstSeen": e.get("firstSeen"), "lastSeen": e.get("lastSeen"),
        "chunkIds": e.get("chunkIds", []), "series": e.get("series", []),
    }


def public_relation(r: dict) -> dict:
    return {
        "id": str(r.get("_id", "")), "from": r["from"], "to": r["to"], "kind": r["kind"],
        "evidence": r.get("evidence", ""), "chunkId": r.get("chunkId"), "date": r.get("date"),
    }
